keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+    ReplicasToDismiss int32 // name TBD
+    ...
+}
+```
+
+```golang
+type DeploymentStatus struct {
+    ...
+    ReplicasToDismiss int32 // name TBD
+    ...
+}
+```
+
+### Disruption Controller: Eviction Protection and Observability
+
+With the `.spec.MinAvailable` and `.spec.MaxUnavailable` options, PDB could allow evicting the surge
+pods because `.status.CurrentHealthy` could become higher than `.status.desiredHealthy`. This could
+disrupt ongoing migration of an application. To support mission-critical applications, we would add
+a `.spec.maxSurge` field. This would cause the disruption controller to count the
+`.status.desiredHealthy` and `.status.DisruptionsAllowed` with max surge in mind. This field could
+serve two purposes:
+
+1. If set to true, it could change the counting of status fields based on the pod workload
+   maxSurge (deployments) or based on the Surge API (see below)
+2. If set to a number (int or percentage), it would override the detected maxSurge and use only the
+   maxSurge value in the PDB object.
+
+
+```golang
+type PodDisruptionBudgetSpec struct {
+    ...
+    // if true, prevents disruption of all detected surge pods
+    // if integer or percentage, prevents disruption of the defined amount of surge pods
+    MaxSurge int32|string|bool // can we do BoolOrIntOrString ?
+    ...
+}
+
+type PodDisruptionBudgetStatus struct {
+    ...
+    // total number of pods that are expected to surge during a node maintenance
+    ExpectedSurgePods int32
+    ...
+}
+```
+
+
+The disruption controller should check whether a workload guarded by a PDB has a pod surge capability.
+If it does, it should report `.status.expectedSurgePods` to allow cluster admins to assess which PDBs
+are blocking cluster upgrade. If the `.status.disruptionsAllowed` equals to 0, because for example
+`.spec.maxUnavailable` is equal to 0, the cluster admin can assess that the PDB will not block
+cluster upgrade if he or she observes `.status.expectedSurgePods` greater than 0.
+
+
+### Surge API (Optional)
+
+Any CustomResourceDefinition or built-in type could implement a surge API. This could be a
+subresource similar to a scale API, that would serve for advertising the surge support. If a
+certain workload supports a pod surge, then one of `.status.availableSurgeReplicas` or
+`.status.currentSurgeReplicas` values will be positive.
+
+```golang
+type Surge struct {
+    ...
+    Status SurgeStatus
+}
+
+type SurgeStatus struct {
+    // number of replicas that can surge during an application rollout or node maintenance
+    SurgeReplicas int32
+    // current number of surge replicas that are running above the normal amount of replicas
+    CurrentSurgeReplicas int32
+}
+
+```
+
+This API would be then used by kubectl and disruption controller to reason about the capabilities
+of an application. It would be similar to how we use the scale API in podControllerFinders in
+disruption controller.
```

Comment:
Would the Surge API be primarily for stateless surge? One other key controller to think about here is StatefulSet. During updates StatefulSet can respect a `maxUnavailable` field: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/961-maxunavailable-for-statefulset. This effectively controls the velocity of the update. MaxSurge achieves the same thing, but with surge rather than unavailability.

The other complexity is around pod deletion ordering. For StatefulSet, this may matter. Rather than having a Surge API that reads the maxSurge parameter and delete pods, could the disruption controller just defer to the operator's standard pod update procedure? This may not work as well in the case where there is only a partial maintenance event (in this case the operator may want to optimize which pods need to be updated). The benefit is that you don't need to re-declare the maxSurge parameter on a Surge API, you can just have a passthrough of responsibility of which pods to delete and in which order to the controller/operator.

Reply:
We will probably scrap the Surge API  and start simpler. We can revisit it later if needed. I will update the KEP to reflect that.

Reply:
I have removed the maxSurge API.